### PR TITLE
ci: ensure roaming between networks doesn't abort file download

### PR DESCRIPTION
**Pull request by thomaseizinger**
This adds an integration test that downloads a 10MB file from a server and simulates the client roaming to another network while the download is active.

We use a DNS resource for this to ensure it also doesn't take too long in that case. DNS resources are what most users will be using and we clear some internal DNS caches on connection failures. Hence, using a DNS resource here is a somewhat roundabout way to test that we aren't failing and re-establishing the connection but migrate it to a new network path.

### CODE REVIEW

**Comment by thomaseizinger** [comment]
To reproduce the issue:

- Start the docker compose setup
- **Disconnect** both client and gateway from the 2nd network:
```
sudo docker network disconnect firezone_app2 firezone-client-1
sudo docker network disconnect firezone_app2 firezone-gateway-1
```
- Start pinging a resource:
```
sudo docker compose exec -it client sh -c 'until ping 172.20.0.100; do true; done'
```
- Switch networks and signal to the linux client to reload configuration:
```
sudo docker network connect firezone_app2 firezone-client-1
sudo docker network connect firezone_app2 firezone-gateway-1
sudo docker network disconnect firezone_app firezone-client-1
sudo kill -s HUP $(ps -C firezone-linux-client -o pid=)
```

When viewing the logs of the client, you'll see that it times out.

Currently, this test setup doesn't fully work yet because we only get returned 1 relay for the initial connection which makes it impossible to discover our srflx address after the network switch.

**Comment by thomaseizinger** [comment]
@AndrewDryga I test this network switching in docker compose, I had to add a 2nd relay. I noticed I still only get 1 relay returned from the portal for the initial connection. Is that because I am using the exact same token? How do I get a 2nd token for this relay?

**Comment by thomaseizinger** [comment]
I am now taking a different approach. I'll be porting some stuff from the snownet-tests to the main docker compose file, in particular a dedicated network for the client with a router. That means the relay's IP can remain stable and we can just disconnect the client from one network and connect to it to the other which will give it a different local and srflx IP. That setup is much closer to production as well so I think that is a good change.

**Comment by thomaseizinger** [comment]
Okay, I managed to reproduce the problem. The command to simulate the roaming is:

```
sudo docker network disconnect firezone_app firezone-client-1
sudo docker network connect firezone_app firezone-client-1 --ip 172.28.0.200
sudo kill -s HUP $(ps -C firezone-linux-client -o pid=)
```

I think we can incorporate this into CI as well!

**Comment by thomaseizinger** [comment]
Given that the primary fix was merged in https://github.com/firezone/firezone/pull/4242, I am going to repurpose this PR to add a test to CI for roaming.

**Comment by thomaseizinger** [comment]
Interesting, this now seems to reproduce this "Pending connection" bug we saw the other day @conectado.

File: docker-compose.yml [hunk 1]
```diff
@@ -357,6 +357,19 @@ services:
       resources:
         ipv4_address: 172.20.0.100
 
+  download.httpbin: # Named after `httpbin` because that is how DNS resources are configured for the test setup.
+    build:
+      target: dev
+      context: rust
+      dockerfile: Dockerfile
+      args:
+        PACKAGE: http-test-server
```
Comment (thomaseizinger) [inline docker-compose.yml]: Need to decide whether we want to build this on-demand every time we run the tests or build is part of CI? Maybe it should not even be part of the mono repo? I don't know :shrug:

**Comment by thomaseizinger** [comment]
@jamilbk This PR highlights a limitation of the CI setup of building everything for `base` too: When we add new packages, those checks will be red. I think that is okay though :)

**Comment by thomaseizinger** [comment]
Interestingly, the compatibility tests pass too :joy: 

This makes me question how useful this test actually is if we don't even seem to need to call `reconnect` for the Linux client? I guess what ends up happening is that we create peer reflexive candidates and that updates the socket? I am a bit confused.

**Review by ReactorScram — APPROVED** [review]
LGTM

**Comment by jamilbk** [comment]
> When we add new packages, those checks will be red. I think that is okay though :)

Yeah, a downside, but those tests aren't required for merge I believe so shouldn't block this.

**Comment by thomaseizinger** [comment]
@jamilbk @conectado Can you review this and let me know whether it is a useful addition?

**Review by jamilbk — COMMENTED** [review]
I think the concept is good, but I'm not sure it makes sense to write a custom server for this? That reduces the number of engineers who can maintain and debug this test, and it also means we need to be sure the custom server included here is free of bugs as well.

Would it be a good idea to use a common HTTP server for this like nginx or caddy? Might also catch issues with actual nginx in the wild that this custom server doesn't trigger.

tl;dr -- this tests that we can download a file from our own custom server but it might be more valuable to test against a popular HTTP server

**Comment by ReactorScram** [comment]
Yeah even `python -m http.server` (or whatever that's called) could work, it's already in Docker so we can pull that image, and it's not a perf test. If Caddy can match that easy default config, Caddy's nice too.

**Review by conectado — APPROVED** [review]
Personally, I find it clearer the rust code than another service

**Comment by thomaseizinger** [comment]
> I think the concept is good, but I'm not sure it makes sense to write a custom server for this? That reduces the number of engineers who can maintain and debug this test, and it also means we need to be sure the custom server included here is free of bugs as well.
> 
> Would it be a good idea to use a common HTTP server for this like nginx or caddy? Might also catch issues with actual nginx in the wild that this custom server doesn't trigger.
> 
> tl;dr -- this tests that we can download a file from our own custom server but it might be more valuable to test against a popular HTTP server

I am open to solutions but there are a few questions:

1. Have you looked at the server code? It is trivial. 
2. If we use a file-based HTTP server then we need to store a binary blob of 0s in the GitHub repository / docker container which seems like a hack.
3. Having control over this server as part of the mono-repo seems easier to maintain. Perhaps we want to test other kinds of high-traffic connections at some point? Or file-uploads?

**Comment by ReactorScram** [comment]
tbf we could use `pv` or `dd` to generate 10 MB of zeroes, it doesn't have to be checked in to the repo.

But yeah... then we'd have to mount it to a container and stuff, so this could be better... I'm surprised `httpbin` doesn't just have a knob to allow very large downloads?

**Comment by thomaseizinger** [comment]
> But yeah... then we'd have to mount it to a container and stuff, so this could be better... I'm surprised `httpbin` doesn't just have a knob to allow very large downloads?

I was surprised too. It is capped at a low number, I think it was 1 MB. I tried several forks of it too and they were all similarly clunky to use so I just gave up and wrote my own in like 2 minutes :)

The only think that bothers me about this is that it increases our CI jobs because we have to build it the same as we build other images. But then I also didn't want to move it out of the mono-repo right away because we may want to iterate on it. If we find it to be very stable, it would a good candidate though!

**Review by jamilbk — APPROVED** [review]
Alrighty, it might be nice being able to "fuzz" an http server for our purposes.
